### PR TITLE
Decorator

### PR DESCRIPTION
**Pull request by MICHAELKITH**
**Description**

- [ ] This pull request adds the implementation of the Decorator design pattern to correct and modify people's names in the Ruby project. It introduces three new classes: Nameable, BaseDecorator, CapitalizeDecorator, and TrimmerDecorator. These classes follow best practices for Ruby development and adhere to the project requirements.

**Changes Made**

- [ ] Created the Nameable class, which serves as an interface and defines the correct_name method. This method raises a NotImplementedError to enforce implementation in derived classes.

- [ ] Modified the Person class to inherit from Nameable, ensuring that the correct_name method is implemented in the Person class. The implementation of correct_name in the Person class simply returns the name attribute.

- [ ] Introduced the BaseDecorator class, which inherits from Nameable. This class acts as a base decorator and contains an instance variable @nameable to hold the object being decorated. The correct_name method of BaseDecorator returns the result of calling correct_name on the @nameable object.

- [ ] Implemented the CapitalizeDecorator class, which inherits from BaseDecorator. This class overrides the correct_name method and capitalizes the output of @nameable.correct_name.

- [ ] Implemented the TrimmerDecorator class, which also inherits from BaseDecorator. This class overrides the correct_name method and ensures that the output of @nameable.correct_name has a maximum of 10 characters. If the output is longer, it trims the name.

**Additional Changes**

- [ ] Ensured that there are no linter errors by running static code analysis tools and addressing any reported issues.

- [ ] Followed the correct Gitflow by creating a new feature branch for this implementation and maintaining a clean commit history.

- [ ] Documented the changes made in a professional manner, including the purpose of each class, the changes to existing classes, and the implementation details of each decorator.

- [ ] Checked the usage of the Decorator pattern in Ruby and ensured that the implemented solution adheres to established best practices for Ruby development.

### CODE REVIEW

File: person.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  def of_age?
+    @age >= 18
+  end
+  private :of_age?
+
+  def can_use_services?
+    of_age? || @parent_permission
+  end
+  public :can_use_services?
+end
```
Comment (Ibizugbe) [inline person.rb]: Nice job creating the `Nameable` class and implementing the required features 👍. Just as required in this milestone, kindly ensure your `person` class inherits the properties of the `Nameable`👍. Also, ensure the class has a method `correct_name`  that returns the name attribute as required👍. 
image

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @MICHAELKITH 👋👋👋, 

tryagain


Good job so far!😎
There are some issues that you still need to work on to go to the next project but you are almost there!👍

## Highlights✅✅

- [x] Linters are passing👍.
- [x] A `Nameable` class with a method `correct_name` have been implemented👍.
- [x] The `Decorator` class inherits from `Nameable`👍.
- [x] The `TrimmerDecorator` and `CapitalizeDecorator` class have been properly implemented👍.

## Required Changes ♻️

- [ ] Nice job documenting your project on your readme👍. Your MIT license link is currently broken, kindly fo well to fix it as this can give your repo a less professional look🥲.
image


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md [hunk 2]
```diff
@@ -1 +1,184 @@
...
+
+I would like to thank Microverse which provided me all knowledge about full stack development.
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- LICENSE -->
+
+## 📝 License <a name="license"></a>
+
+This project is [MIT](./MIT) licensed.
```
Comment (Qoosim) [inline README.md]: - [ ] Your MIT License link is yet to be fixed. It was mentioned by the previous reviewer, but yet to be fixed. Kindly fix it.

license

File: person.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
```
Comment (Qoosim) [inline person.rb]: - [ ] As mentioned by the previous reviewer, just as required in this milestone, kindly ensure your person class inherits the properties of the Nameable. 
- [ ] Also, ensure the Person class has a method correct_name that returns the name attribute as required.

### Project Requirements
project-rqt

**Review by Qoosim — CHANGES_REQUESTED** [review]
#  *Status: Required Changes* :recycle: 

Hi @MICHAELKITH, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


## Highlights
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark:  



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

- [OPTIONAL] You were actually following wrong workflow. You're required to use [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md), not GitHub Flow. Has this not escaped the watch of the previous reviewer, your PR was supposed to be invalidated. Kindly watch out for this in every project you build at Microverse before starting your project at all.

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 2 above)
Comment (KDlamini) [reply README.md]: Well Resolved ✔️ 

image

File: person.rb (hunk 3 above)
Comment (KDlamini) [reply person.rb]: Not Resolved ✖️ 

image

File: person.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+require './nameable'
+
+class Person
+  attr_reader :id
+  attr_accessor :name, :age
```
Comment (KDlamini) [inline person.rb]: As requested by the previous reviewers, kindly make sure that the `Person` class inherits from the `Nameable` class.

This is how to implement inheritance:
```ruby
class Person < Nameable
  .
  .

    def correct_name
       name
    end
  .
  .
end
```

**Review by KDlamini — CHANGES_REQUESTED** [review]
### Hi @MICHAELKITH 🙋‍♀️,  

You have done an incredible job implementing some of the changes requested by the previous reviewer. 👏🏾👏🏾👏🏾  
However, there are still some changes that aren't addressed yet. You are almost there! 💪🏾 

## Highlights 🌞

✔️ Good PR title and description
✔️ The Nameable class has a method correct_name implemented.
✔️ The Decorator class inherits from Nameable.
✔️ The TrimmerDecorator and CapitalizeDecorator classes are present.

## Required Changes ♻️
 
**Kindly look through the comments below 👇 for required changes**

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me @KDlamini in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by KDlamini — APPROVED** [review]
### Hi @MICHAELKITH 🙋‍♀️,  

Excellent work! 💯 
You have done a great job implementing all the required changes. 🎊 🥳 
Your feature is complete! There is nothing else to say other than... it's time to merge it. :shipit: 
Congratulations!!!

## HIGHLIGHTS 🌟 

✔️ Good PR title and description
✔️ The Nameable class has a method correct_name implemented.
✔️ The Person class inherits from Nameable.
✔️ The Decorator class inherits from Nameable.
✔️ The TrimmerDecorator and CapitalizeDecorator classes are present.

# Status: Approved ✔️ 

oscars-standing-ovation

Cheers 🥂  and Happy coding!!! 👯 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me @KDlamini  in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**